src/lib.rs

```diff
@@ -568,63 +782,56 @@ impl<T> Receiver<T> {
+                // our write of the waker. We use relaxed ordering on failure since the sender does
+                // not need to syncrhonize with our write and the individual match arms handle any
+                // additional synchronization
                 match channel
                     .state
-                    .compare_exchange(EMPTY, RECEIVING, SeqCst, SeqCst)
+                    .compare_exchange(EMPTY, RECEIVING, Release, Relaxed)
                 {
-                    // We stored our waker, now we park until the sender has changed the state
-                    Ok(EMPTY) => loop {
-                        let (state, timed_out) = if let Some(timeout) =
-                            deadline.checked_duration_since(Instant::now())
-                        {
-                            thread::park_timeout(timeout);
-                            (channel.state.load(SeqCst), false)
-                        } else {
-                            // We reached the deadline. Stop being in the receiving state.
-                            (channel.state.swap(EMPTY, SeqCst), true)
-                        };
-                        match state {
-                            // The sender sent the message while we were parked.
-                            MESSAGE => {
-                                channel.state.store(DISCONNECTED, SeqCst);
-                                break Ok(unsafe { channel.take_message() });
-                            }
-                            // The sender was dropped while we were parked.
-                            DISCONNECTED => break Err(RecvTimeoutError::Disconnected),
-                            // State did not change, spurious wakeup, park again.
-                            RECEIVING => {
-                                if timed_out {
-                                    unsafe { channel.drop_waker() };
-                                    break Err(RecvTimeoutError::Timeout);
-                                }
-                            }
-                            _ => unreachable!(),
-                        }
-                    },
-                    // The sender sent the message while we prepared to park.
+                    // We stored our waker, now we delegate to the callback to finish the receive
+                    // operation
+                    Ok(EMPTY) => finish(channel),
+                    // The sender sent the message while we prepared to finish
                     Err(MESSAGE) => {
-                        channel.state.store(DISCONNECTED, SeqCst);
+                        // See comments in `recv` for ordering and safety
+
+                        fence(Acquire);
+
                         unsafe { channel.drop_waker() };
+
+                        // ORDERING: the sender has been `mem::forget`-ed so this update only
+                        // needs to be visible to us
+                        channel.state.store(DISCONNECTED, Relaxed);
```

Comment:
We're allowed to assume that the process of transferring ownership between threads has appropriate acquire/release semantics. So in this particular case that store would be released, and the acquired by the other thread, providing sufficient synchronization.

Reply:
Ok. I was thinking along those lines as well and assumed that would be the case for most/all common ways of sending the instance over. And I could not come up with any real way of proving otherwise.